### PR TITLE
Optimize parser

### PR DESCRIPTION
**Pull request by mpenick**
This replaces the existing antlr4 parser with a recursive descent parser
that uses Ragel for the lexer implementation. This results in a
*significant* speed up. A rough benchmark w/ a non-optimal case shows a
~7000x speed up:

ParserAntlr: 187771600 ns/op 0.121691 mib/s
ParserRagel: 24835 ns/op 920.041225 mib/s

A more realistic benchmark wouldn't have as many "handled" queries and
would be faster for the Ragel-based parser.

Currently, this only detects "SELECT" queries as idempotent. A follow up
patch will add implementations for other query types.

### CODE REVIEW

**Comment by mpenick** [comment]
Benchmark is here: https://github.com/datastax/cql-proxy/blob/optimize-parsing-full/parser/bench/bench.go

**Comment by mpenick** [comment]
*Note: `lexer.go` can be ignored, it's generated code (from `lexer.rl`)

File: parser/lexer_test.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	}
+
+	for _, tt := range tests {
+		var l lexer
+		l.init(tt.literal)
+		assert.Equal(t, tt.tk, l.next(), fmt.Sprintf("failed on literal: %s", tt.literal))
+	}
+}
+
+func TestLexerFloat(t *testing.T) {
```
Comment (dougwettlaufer) [inline parser/lexer_test.go:77]: Why is this called "float" but it's testing integers?

File: parser/parse_select.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parser
+
+import (
+	"errors"
+)
+
+// Determines is the proxy handles the select statement.
```
Comment (dougwettlaufer) [inline parser/parse_select.go:21]: Handled here means it's one of the system or peers tables?

File: parser/parser_utils.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+		}
+	}
+	return false
+}
+
+func columnFromSelector(selector Selector, columns []*message.ColumnMetadata, table string) (column *message.ColumnMetadata, err error) {
+	switch s := selector.(type) {
+	case *CountStarSelector:
+		return &message.ColumnMetadata{
+			Keyspace: "system",
```
Comment (dougwettlaufer) [inline parser/parser_utils.go:105]: I'm probably missing something but why is this always `system` keyspace?

**Review by dougwettlaufer — APPROVED** [review]

File: parser/lexer_test.go (hunk 1 above)
Comment (mpenick) [reply parser/lexer_test.go:77]: Good catch. Dead code.

File: parser/parser_utils.go (hunk 3 above)
Comment (mpenick) [reply parser/parser_utils.go:105]: Currently, all handled queries are in the "system" keyspace. I agree this doesn't really belong here. I've added `Keyspace` to the `SelectStatement` type so this decision isn't made here anymore.

File: parser/parse_select.go (hunk 2 above)
Comment (mpenick) [reply parser/parse_select.go:21]: It's always tables in "system", but could be "local", "peers", as well as legacy schema tables. I've added a comment to clarify this.

**Pull request merged**